### PR TITLE
replace pandas dataframe with spark dataframe in data quality signal

### PR DESCRIPTION
**Pull request by VivienneTang**
replace pandas dataframe with spark dataframe in data quality signal

### CODE REVIEW

File: assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py [hunk 1]
```diff
@@ -542,7 +551,7 @@ def compute_data_quality_metrics(df, data_stats_table):
         .when(col("dataType") == "DateType()", "Categorical")
         .when(col("dataType") == "FloatType()", "Numerical")
         .when(col("dataType") == "ShortType()", "Numerical")
-        .when(col("dataType") == "ByteType()", "Categorical")
+        .when(col("dataType") == "ByteType()", "Numerical")
```
Comment (ycheng35xo) [inline assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py:554]: why making this change? When I checked with Alex he mentioned that Binary should be Categorical

File: assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py [hunk 2]
```diff
@@ -79,19 +79,28 @@ def get_df_schema(df: pyspark.sql.DataFrame) -> pyspark.sql.DataFrame:
...
 
     Returns:
-        na_metric_df: A Pypsark Pandas DataFrame containing the number of null values
+        na_metric_df: A Pypsark DataFrame containing the number of null values
     for each column of the input PySpark DataFrame.
     """
-    na_metric_df = ps.DataFrame(df.isna().sum()).reset_index()
-    na_metric_df.columns = ["featureName", "violationCount"]
+    na_metric_df = df.select([count(when(isnan(c) | col(c).isNull(), c)).alias(c) for c in df.columns])
+    na_metric_df.show()
```
Comment (ycheng35xo) [inline assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py:94]: remove this?

File: assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py (hunk 1 above)
Comment (VivienneTang) [reply assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py:554]: This is not Binary, this is byte type, when I recheck byte type in spark, it is numerical.

File: assets/model_monitoring/components/src/data_quality_statistics/run.py [hunk 3]
```diff
@@ -22,9 +22,7 @@ def run():
 
     metric_unique_df = compute_data_quality_statistics(df)
 
-    sp_metric_unique_df = (
-        metric_unique_df.to_spark()
-    )  # Convert back to Spark dataframe to output as MLTable
+    sp_metric_unique_df = metric_unique_df
```
Comment (ycheng35xo) [inline assets/model_monitoring/components/src/data_quality_statistics/run.py:25]: no need to have this equal? Just rename?

File: assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py [hunk 4]
```diff
@@ -79,19 +79,28 @@ def get_df_schema(df: pyspark.sql.DataFrame) -> pyspark.sql.DataFrame:
...
 
-def get_null_count(df: ps.DataFrame) -> ps.DataFrame:
+def get_null_count(df: pyspark.sql.DataFrame) -> pyspark.sql.DataFrame:
     """
-    Compute a Pandas DataFrame containing the number of null values for each column of the input PySpark DataFrame.
+    Compute a PySpark DataFrame containing the number of null values for each column of the input PySpark DataFrame.
 
     Args:
-        df: Input PySpark Pandas DataFrame.
+        df: Input PySpark  DataFrame.
```
Comment (ycheng35xo) [inline assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py:87]: nit: extra space?

File: assets/model_monitoring/components/src/data_quality_statistics/compute_data_quality_statistics.py [hunk 5]
```diff
@@ -111,65 +111,77 @@ def get_unique_value_list(df: pyspark.sql.DataFrame) -> ps.DataFrame:
...
     Compute the maximum value for each feature in the input DataFrame.
 
     Args:
-        df: A pandas-on-Spark DataFrame.
+        df: A spark data frame with features to calculate max value and min value。
+        dtype_df: A spark data frame for datatype for each featureName
 
     Returns:
-        A pandas-on-Spark DataFrame containing two columns: "featureName" and "max_value".
+        A Spark DataFrame containing two columns: "featureName" and "max_value".
```
Comment (ycheng35xo) [inline assets/model_monitoring/components/src/data_quality_statistics/compute_data_quality_statistics.py:126]: This comment should be changed to "max_value" and "min_value"

Comment (ycheng35xo) [reply assets/model_monitoring/components/src/data_quality_statistics/compute_data_quality_statistics.py:126]: Also, line 128

File: assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py (hunk 4 above)
Comment (VivienneTang) [reply assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py]: removed

File: assets/model_monitoring/components/src/data_quality_statistics/run.py (hunk 3 above)
Comment (VivienneTang) [reply assets/model_monitoring/components/src/data_quality_statistics/run.py]: removed

File: assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py (hunk 2 above)
Comment (VivienneTang) [reply assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py]: removed

File: assets/model_monitoring/components/src/data_quality_statistics/compute_data_quality_statistics.py (hunk 5 above)
Comment (VivienneTang) [reply assets/model_monitoring/components/src/data_quality_statistics/compute_data_quality_statistics.py]: fixed the docstring

File: assets/model_monitoring/components/src/data_quality_statistics/compute_data_quality_statistics.py [hunk 6]
```diff
@@ -6,16 +6,16 @@
...
 import pyspark
-import pyspark.pandas as ps
 
 
 # Init spark session
 sc = SparkContext.getOrCreate()
 spark = SparkSession(sc)
 
-supported_datatype_for_max_min_value = ["IntegerType()", "DoubleType()", "ByteType()", "LongType()", "FloatType()"]
+supported_datatype_for_max_min_value = ["IntegerType()", "DoubleType()", "ByteType()",
```
Comment (ycheng35xo) [inline assets/model_monitoring/components/src/data_quality_statistics/compute_data_quality_statistics.py:17]: I think our PR has some dependencies on each other: 

For example, this list and the function  get_features_for_max_min_calculation can be replaced with the numerical_columns from my PR

**Review by ycheng35xo — DISMISSED** [review]

**Review by ycheng35xo — APPROVED** [review]

**Review by ycheng35xo — APPROVED** [review]

**Pull request merged**